increase Hourglasses and gemCapExtra promptly when multi-month subscription renews - fixes #4819

Alys opened the pull request:
Fixes #4819

The bug in #4819 is that when a multi-month recurring subscription renews, the user does not immediately get the new Hourglasses and increased gem cap that they've paid for. Instead, as each month after that passes, they build up credit towards one new Hourglass in the same way that a one-month recurring subscription does. This is unfair since they've already paid for the full multi-month period.

This PR fixes that, mostly. We can't increment the Hourglasses and extra gem cap at the exact date that the subscription renews because we don't track payments to the extent needed for that. 
Instead, this PR waits until the start of the month immediately after the subscription has renewed and then gives the extra Hourglasses and gem cap.

It's not a perfect solution because there's already a lot of subscriptions that have completed their first period and are part way through a second/third/etc period. For example, consider a 6-month subscription that's four months into its second year. When that user crons at the start of the next month after this PR goes live, they'll be given two new Hourglasses and increased gem cap, even though they actually got one of those new Hourglasses already at the end of the previous three-month period - i.e., we're giving them slightly more than they should have. However many of the subscribers in this situation will have suffered through months of not having the correct benefits in advance, so this balances the scales a bit. In the case of that example, their increased subscription benefits will continue to be offset by four months indefinitely, which isn't ideal but is still to their benefit and if anyone reports the problem we can manually adjust their data easily enough (just as we do for reports that are happening now). I'm considering modifying the code to take into account the current `count` and to adjust the new benefits and offset to match it, but I don't want to make the code too complex and risk introducing new bugs.

Alys committed:
allow Hourglasses and gemCapExtra to increase promptly after a multi-month subscription has renewed

paglias commented on website/server/libs/cron.js:
```diff
@@ -70,16 +71,49 @@ function grantEndOfTheMonthPerks (user, now) {
-        plan.consecutive.trinkets++;
-        plan.consecutive.gemCapExtra += 5;
-        if (plan.consecutive.gemCapExtra > 25) plan.consecutive.gemCapExtra = 25; // cap it at 50 (hard 25 limit + extra 25)
+      plan.consecutive.offset--;
```
why was the check for this to be > 1 removed?

Alys replied:
With the new code, the offset is always decremented, and when it's reached -1, that's a sign that we need to look into special handling for multi-month subscriptions that renewed in the previous month. The comments below this line go into some detail about that but I'm happy to make it clearer if needed!

Alys committed:
Merge remote-tracking branch 'upstream/develop' into hourglasses-fix

Alys committed:
fix existing Hourglass and Gem Cap tests that were wrong

The scenario originally used for these two tests was a six-month recurring
subscription (you can tell that from the starting offset having a non-zero value).
For recurring subscriptions, we do NOT want to increase the consecutive month
benefits as soon as the sixth month starts because the user has already been
given a full six months' benefits in advance and they might cancel the
subscription before it renews later in the sixth month.
Therefore we want to give the extra benefits at the beginning of the seventh
month (ideally we'd give them mid-month in the sixth month when the renewal
happens but we don't have support for tracking renewal dates).
So, the two changed tests were actually not correct for the case
where the offset started as non-zero.

These tests are correct for one-month recurring subscriptions (when the offset
is never set to anything above zero). The user isn't meant to get any consecutive
month benefits until a multiple of 3 months has been reached.

Alys committed:
add tests for one-month recurring subscription before 3x months are reached

Alys committed:
add tests for 3-, 6-, and 12-month recurring subscriptions

The 3-month tests are the most thorough, stepping through the
expected start and end values of consecutive data for a 7-month
range.

The 6-month tests are a bit less thorough since the same code is
used for all multi-month periods.
The discount Google subscription code is used to ensure we keep
support for it.

The 12-month tests are less thorough still, since again the same
code is used.

I'm about to try some more tests with `useFakeTimers`, which should
be a better way to test the code since they won't rely on me having
set the initial values correctly for each test. :) But I wanted to
work through these cases manually first to ensure my understanding
of how the values should change does actually match the code.

Alys committed:
add tests for 1-, 3-, 6-, and 12-month recurring subscriptions using clock changes to simulate passing months

Also fixed the clock call in an unrelated test because it was forming
the date incorrectly (`unix()` can't be used to create a date).

Also changed email@email.email to email@example.com because
email@email.email is potentially a real email address.

paglias commented:
Is this ready for review @Alys ?

Alys commented:
No, I'm still adding tests. I'll comment and change the labels when I'm
done.

Alys committed:
add tests for 3-month gift subscriptions - no extra consecutive benefits given

Alys committed:
add tests for consecutive benefits for 6-month recurring subscription that has incorrect consecutive month data because it started before issue #4819 was fixed

Alys commented:
The code and tests are finished and this is ready for review.

Alys committed:
fix lint errors

paglias commented on test/api/v3/unit/libs/cron.test.js:
```diff
@@ -143,15 +144,16 @@ describe('cron', () => {
     });
 
     it('increments plan.consecutive.gemCapExtra multiple times if user has been absent with continuous subscription', () => {
+      // TODO if approved, delete this because the new tests below include this situation
```
what about these comments @Alys ?

Alys replied:
It's up to you. If you're happy with that the new tests cover what these old tests cover, and if you don't want to keep the old tests because they're related to other tests around them, then I can remove them now.

paglias replied:
ah ok, let's delete them then!

Alys replied:
I've deleted those four tests.

paglias reviewed (approved):
This is looking good, just left one comment

Alys committed:
remove outdated subscription tests
